source/Calamari/Kubernetes/ManifestReporter.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.ServiceMessages;
+using Calamari.Common.Plumbing.Variables;
+using YamlDotNet.Core;
+using YamlDotNet.RepresentationModel;
+using YamlDotNet.Serialization;
+using YamlDotNet.Serialization.NamingConventions;
+
+namespace Calamari.Kubernetes
+{
+    public interface IManifestReporter
+    {
+        void ReportManifestApplied(string filePath);
+    }
+
+    public class ManifestReporter
+    {
+        readonly IVariables variables;
+        readonly ICalamariFileSystem fileSystem;
+        readonly ILog log;
+
+        static readonly ISerializer YamlSerializer = new SerializerBuilder()
+                                                     .WithNamingConvention(CamelCaseNamingConvention.Instance)
+                                                     .Build();
+
+        public ManifestReporter(IVariables variables, ICalamariFileSystem fileSystem, ILog log)
+        {
+            this.variables = variables;
+            this.fileSystem = fileSystem;
+            this.log = log;
+        }
+
+        string GetNamespace(YamlMappingNode yamlRoot)
+        {
+            var implicitNamespace = variables.Get(SpecialVariables.Namespace) ?? "default";
+            
+            if (yamlRoot.Children.TryGetValue("metadata", out var metadataNode) && metadataNode is YamlMappingNode metadataMappingNode &&
+                metadataMappingNode.Children.TryGetValue("namespace", out var namespaceNode) && namespaceNode is YamlScalarNode namespaceScalarNode &&
+                !string.IsNullOrWhiteSpace(namespaceScalarNode.Value))
+            {
+                implicitNamespace = namespaceScalarNode.Value;
+            }
+
+            return implicitNamespace;
+        }
+
+        public void ReportManifestApplied(string filePath)
+        {
+            using (var yamlFile = fileSystem.OpenFile(filePath, FileAccess.ReadWrite))
+            {
+                try
+                {
+                    var yamlStream = new YamlStream();
+                    yamlStream.Load(new StreamReader(yamlFile));
+
+                    foreach (var document in yamlStream.Documents)
+                    {
+                        if (!(document.RootNode is YamlMappingNode rootNode))
+                        {
+                            continue;
+                        }
+
+                        var updatedDocument = YamlSerializer.Serialize(rootNode);
+                        
+                        var ns = GetNamespace(rootNode);
```

Comment:
Getting the namespace on this side means that we dont need to load anything on the server side.
Unlike things like space\project etc, its not particularly sensitive since we would have otherwise trusted what was passed over in the manifest as the default anyway.